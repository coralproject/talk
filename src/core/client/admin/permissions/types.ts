import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
