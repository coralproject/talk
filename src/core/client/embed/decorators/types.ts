import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
