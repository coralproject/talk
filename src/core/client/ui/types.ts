import React from "react";

// TODO: Extract useful common types into its own package.

/**
 * Overwrite properties of `T`.
 *
 * E.g. Omit<{a: boolean, b: boolean}, "b"> = {a: boolean}
 */
export type Omit<U, K extends keyof U> = Pick<U, Exclude<keyof U, K>>;

/**
 * Overwrite properties of `T`.
 *
 * E.g. Overwrite<{a: boolean}, {a: string}> = {a: string}
 */
export type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> & U;

/**
 * Returns the PropTypes from a React Component.
 *
 * E.g. type ButtonProps = ReturnPropTypes<Button>;
 */
export type PropTypesOf<T> = T extends React.ComponentType<infer R>
  ? R
  : T extends React.Component<infer S> ? S : {};
