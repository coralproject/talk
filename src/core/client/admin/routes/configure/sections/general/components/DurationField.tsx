import { Localized } from "fluent-react/compat";
import React, { ChangeEvent, Component } from "react";

import { Flex, Option, SelectField, TextField } from "talk-ui/components";

import styles from "./DurationField.css";

interface Props {
  name: string;
  value: number;
  disabled: boolean;
  onChange: (v: number) => void;
}

interface State {
  value?: number;
  unit?: number;
}

const SECONDS_IN_AN_HOUR = 3600;
const SECONDS_IN_A_DAY = 86400;
const SECONDS_IN_A_WEEK = 604800;

function valueToState(value: number, unit?: number) {
  if (!unit) {
    unit = SECONDS_IN_AN_HOUR;
    if (value > 0) {
      if (value % SECONDS_IN_A_WEEK === 0) {
        unit = SECONDS_IN_A_WEEK;
      } else if (value % SECONDS_IN_A_DAY === 0) {
        unit = SECONDS_IN_A_DAY;
      }
    }
  }
  return { unit, value: value / unit };
}

function stateToValue(state: State) {
  return state.value! * state.unit!;
}

class DurationField extends Component<Props, State> {
  public state: State = this.props.value ? valueToState(this.props.value) : {};
  public componentWillReceiveProps(nextProps: Props) {
    this.setState(valueToState(nextProps.value, this.state.unit));
  }
  private handleValueChange = (e: ChangeEvent<HTMLInputElement>) => {
    const newState: State = {
      ...this.state,
      value: !e.target.value ? 0 : parseInt(e.target.value, 10),
    };
    if (this.props.onChange) {
      this.props.onChange(stateToValue(newState));
    }
  };
  private handleUnitChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const newState: State = {
      ...this.state,
      unit: parseInt(e.target.value, 10),
    };
    this.setState(newState, () => {
      if (this.props.onChange) {
        this.props.onChange(stateToValue(newState));
      }
    });
  };

  public render() {
    const { disabled, name } = this.props;
    return (
      <Flex itemGutter>
        <TextField
          className={styles.value}
          name={`${name}-value`}
          onChange={this.handleValueChange}
          value={this.state.value ? this.state.value.toString() : ""}
          disabled={disabled}
          autoComplete="off"
          autoCorrect="off"
          autoCapitalize="off"
          spellCheck={false}
          textAlignCenter
          aria-label="value"
        />
        <SelectField
          name={`${name}-unit`}
          onChange={this.handleUnitChange}
          disabled={disabled}
          aria-label="unit"
          classes={{
            select: styles.unit,
          }}
          value={
            this.state.unit
              ? this.state.unit.toString()
              : SECONDS_IN_AN_HOUR.toString()
          }
        >
          <Localized
            id="configure-general-durationField-hours"
            $value={this.state.value}
          >
            <Option value={SECONDS_IN_AN_HOUR.toString()}>Hours</Option>
          </Localized>
          <Localized
            id="configure-general-durationField-days"
            $value={this.state.value}
          >
            <Option value={SECONDS_IN_A_DAY.toString()}>Days</Option>
          </Localized>
          <Localized
            id="configure-general-durationField-weeks"
            $value={this.state.value}
          >
            <Option value={SECONDS_IN_A_WEEK.toString()}>Weeks</Option>
          </Localized>
        </SelectField>
      </Flex>
    );
  }
}

export default DurationField;
