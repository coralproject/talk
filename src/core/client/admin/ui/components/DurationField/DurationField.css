.value {
  width: calc(6 * var(--mini-unit));
}

.select {
  height: 100%;
  min-width: calc(17 * var(--mini-unit));
}

.unit {
  line-height: 36px;
}
