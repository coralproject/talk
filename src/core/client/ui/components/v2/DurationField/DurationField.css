.value {
  width: calc(6 * var(--mini-unit));
}

.select {
  height: 100%;
  min-width: calc(17 * var(--mini-unit));
}

.unit {
  font-family: var(--font-family-primary);
  font-weight: var(--font-weight-primary-semi-bold);
  font-size: var(--font-size-2);
  line-height: 36px;
}
