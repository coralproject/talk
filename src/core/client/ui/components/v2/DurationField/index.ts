export { default, DURATION_UNIT } from "./DurationField";
