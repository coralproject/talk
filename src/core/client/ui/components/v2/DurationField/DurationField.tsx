import { Localized } from "fluent-react/compat";
import React, {
  ChangeEvent,
  FunctionComponent,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import { UNIT } from "coral-common/helpers/i18n";
import { Flex } from "coral-ui/components";

import { Option, SelectField } from "../SelectField";
import TextField, { TextFieldProps } from "../TextField";

import styles from "./DurationField.css";

/**
 * DURATION_UNIT are units that can be used in the
 * DurationField components.
 */
export const DURATION_UNIT = UNIT;

const DURATION_UNIT_MAP = {
  [DURATION_UNIT.SECONDS]: "second",
  [DURATION_UNIT.MINUTES]: "minute",
  [DURATION_UNIT.HOURS]: "hour",
  [DURATION_UNIT.DAYS]: "day",
  [DURATION_UNIT.WEEKS]: "week",
};

interface Props extends Pick<TextFieldProps, "color" | "name" | "disabled"> {
  value: string;
  onChange: (v: string) => void;
  /** Specifiy units to include */
  units?: ReadonlyArray<UNIT>;
}

function convertToSeconds(value: string, unit?: UNIT) {
  const parsed = parseInt(value, 10);
  return (isNaN(parsed) || !unit ? value : parsed * unit).toString();
}

function convertFromSeconds(
  value: string,
  units: ReadonlyArray<UNIT>,
  unit?: UNIT
) {
  const parsed = parseInt(value, 10);

  // If value was a valid number..
  if (!isNaN(parsed)) {
    // If unit is not set, we'll find the best matching unit.
    if (!unit) {
      // Start from the first unit,
      // keep first unit if value is set to 0,
      // otherwise use better matching unit if the value is fully dividable by the unit.
      unit = units.reduce((x, cur) =>
        parsed % cur === 0 && parsed !== 0 ? cur : x
      );
    }
    // Compute new value relative to the selected unit.
    value = (parsed / unit).toString();
  }

  return {
    unit,
    value,
  };
}

/**
 * Duration Field renders a TextField that accepts a number and a SelectField with a unit.
 * If the entered value is a valid number, it'll propagate the computed value in seconds via
 * onChange otherwise it'll just propogate whatever was entered as the value TextField.
 */
const DurationField: FunctionComponent<Props> = ({
  value,
  units = [UNIT.HOURS, UNIT.DAYS, UNIT.WEEKS],
  onChange,
  disabled,
  name,
  color,
}) => {
  const [selectedUnit, setSelectedUnit] = useState(
    convertFromSeconds(value, units).unit
  );

  // If value changes, and selectedUnit has not been set, then set the value.
  useEffect(() => {
    if (!selectedUnit) {
      setSelectedUnit(convertFromSeconds(value, units).unit);
    }
  }, [value]);

  const elementCallbacks = useMemo(() => units.map(k => DURATION_UNIT_MAP[k]), [
    units,
  ]);

  const { value: computedValue } = useMemo(
    () => convertFromSeconds(value, units, selectedUnit),
    [value, selectedUnit]
  );

  const handleValueChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      onChange(convertToSeconds(e.target.value, selectedUnit));
    },
    [onChange, selectedUnit]
  );

  const handleUnitChange = useCallback(
    (e: ChangeEvent<HTMLSelectElement>) => {
      const unit = parseInt(e.target.value, 10);
      setSelectedUnit(unit);
      onChange(convertToSeconds(computedValue, unit));
    },
    [setSelectedUnit, onChange, computedValue]
  );

  return (
    <Flex itemGutter>
      <TextField
        name={`${name}-value`}
        className={styles.value}
        onChange={handleValueChange}
        value={computedValue}
        disabled={disabled}
        autoComplete="off"
        autoCorrect="off"
        autoCapitalize="off"
        spellCheck={false}
        textAlignCenter
        aria-label="value"
        color={color || "regular"}
      />
      {elementCallbacks.length === 1 ? (
        <Localized
          id="framework-durationField-unit"
          $unit={elementCallbacks[0]}
          $value={parseInt(computedValue, 10)}
        >
          <span className={styles.unit}>{elementCallbacks[0]}</span>
        </Localized>
      ) : (
        <SelectField
          name={`${name}-unit`}
          onChange={handleUnitChange}
          disabled={disabled}
          aria-label="unit"
          classes={{
            select: styles.select,
          }}
          value={(selectedUnit || units[0]).toString()}
        >
          {elementCallbacks.map((unit, i) => {
            const val = units[i];
            return (
              <Localized
                id="framework-durationField-unit"
                $unit={unit}
                $value={parseInt(computedValue, 10)}
                key={i}
              >
                <Option value={val.toString()}>{unit}</Option>
              </Localized>
            );
          })}
        </SelectField>
      )}
    </Flex>
  );
};

export default DurationField;
