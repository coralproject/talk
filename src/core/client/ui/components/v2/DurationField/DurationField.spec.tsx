import { noop } from "lodash";
import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { PropTypesOf } from "coral-framework/types";

import DurationField, { DURATION_UNIT } from "./DurationField";

it("renders correctly with default units", () => {
  const props: PropTypesOf<typeof DurationField> = {
    name: "duration",
    value: "",
    disabled: false,
    onChange: noop,
  };
  const renderer = createRenderer();
  renderer.render(<DurationField {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it("renders correctly with specified units", () => {
  const props: PropTypesOf<typeof DurationField> = {
    name: "duration",
    value: "",
    disabled: false,
    onChange: noop,
    units: [DURATION_UNIT.SECONDS, DURATION_UNIT.HOURS],
  };
  const renderer = createRenderer();
  renderer.render(<DurationField {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it("use best matching unit", () => {
  const props: PropTypesOf<typeof DurationField> = {
    name: "duration",
    value: "3600",
    disabled: false,
    onChange: noop,
    units: [DURATION_UNIT.SECONDS, DURATION_UNIT.MINUTES, DURATION_UNIT.HOURS],
  };
  const renderer = createRenderer();
  renderer.render(<DurationField {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it("use initial unit if 0", () => {
  const props: PropTypesOf<typeof DurationField> = {
    name: "duration",
    value: "0",
    disabled: false,
    onChange: noop,
    units: [DURATION_UNIT.SECONDS, DURATION_UNIT.MINUTES, DURATION_UNIT.HOURS],
  };
  const renderer = createRenderer();
  renderer.render(<DurationField {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it("accepts invalid input", () => {
  const props: PropTypesOf<typeof DurationField> = {
    name: "duration",
    value: "this is so invalid",
    disabled: false,
    onChange: noop,
    units: [DURATION_UNIT.SECONDS, DURATION_UNIT.MINUTES, DURATION_UNIT.HOURS],
  };
  const renderer = createRenderer();
  renderer.render(<DurationField {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
