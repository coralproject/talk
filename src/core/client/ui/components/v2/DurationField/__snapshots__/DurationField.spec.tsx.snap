// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts invalid input 1`] = `
<ForwardRef(forwardRef)
  itemGutter={true}
>
  <ForwardRef(forwardRef)
    aria-label="value"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="DurationField-value"
    color="regular"
    disabled={false}
    name="duration-value"
    onChange={[Function]}
    spellCheck={false}
    textAlignCenter={true}
    value="this is so invalid"
  />
  <withPropsOnChange(withUIContext(WithKeyboardFocus))
    aria-label="unit"
    classes={
      Object {
        "select": "DurationField-select",
      }
    }
    disabled={false}
    name="duration-unit"
    onChange={[Function]}
    value="1"
  >
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "second",
          "value": NaN,
        }
      }
    >
      <Option
        value="1"
      >
        second
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "minute",
          "value": NaN,
        }
      }
    >
      <Option
        value="60"
      >
        minute
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "hour",
          "value": NaN,
        }
      }
    >
      <Option
        value="3600"
      >
        hour
      </Option>
    </Localized>
  </withPropsOnChange(withUIContext(WithKeyboardFocus))>
</ForwardRef(forwardRef)>
`;

exports[`renders correctly with default units 1`] = `
<ForwardRef(forwardRef)
  itemGutter={true}
>
  <ForwardRef(forwardRef)
    aria-label="value"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="DurationField-value"
    color="regular"
    disabled={false}
    name="duration-value"
    onChange={[Function]}
    spellCheck={false}
    textAlignCenter={true}
    value=""
  />
  <withPropsOnChange(withUIContext(WithKeyboardFocus))
    aria-label="unit"
    classes={
      Object {
        "select": "DurationField-select",
      }
    }
    disabled={false}
    name="duration-unit"
    onChange={[Function]}
    value="3600"
  >
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "hour",
          "value": NaN,
        }
      }
    >
      <Option
        value="3600"
      >
        hour
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "day",
          "value": NaN,
        }
      }
    >
      <Option
        value="86400"
      >
        day
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "week",
          "value": NaN,
        }
      }
    >
      <Option
        value="604800"
      >
        week
      </Option>
    </Localized>
  </withPropsOnChange(withUIContext(WithKeyboardFocus))>
</ForwardRef(forwardRef)>
`;

exports[`renders correctly with specified units 1`] = `
<ForwardRef(forwardRef)
  itemGutter={true}
>
  <ForwardRef(forwardRef)
    aria-label="value"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="DurationField-value"
    color="regular"
    disabled={false}
    name="duration-value"
    onChange={[Function]}
    spellCheck={false}
    textAlignCenter={true}
    value=""
  />
  <withPropsOnChange(withUIContext(WithKeyboardFocus))
    aria-label="unit"
    classes={
      Object {
        "select": "DurationField-select",
      }
    }
    disabled={false}
    name="duration-unit"
    onChange={[Function]}
    value="1"
  >
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "second",
          "value": NaN,
        }
      }
    >
      <Option
        value="1"
      >
        second
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "hour",
          "value": NaN,
        }
      }
    >
      <Option
        value="3600"
      >
        hour
      </Option>
    </Localized>
  </withPropsOnChange(withUIContext(WithKeyboardFocus))>
</ForwardRef(forwardRef)>
`;

exports[`use best matching unit 1`] = `
<ForwardRef(forwardRef)
  itemGutter={true}
>
  <ForwardRef(forwardRef)
    aria-label="value"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="DurationField-value"
    color="regular"
    disabled={false}
    name="duration-value"
    onChange={[Function]}
    spellCheck={false}
    textAlignCenter={true}
    value="1"
  />
  <withPropsOnChange(withUIContext(WithKeyboardFocus))
    aria-label="unit"
    classes={
      Object {
        "select": "DurationField-select",
      }
    }
    disabled={false}
    name="duration-unit"
    onChange={[Function]}
    value="3600"
  >
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "second",
          "value": 1,
        }
      }
    >
      <Option
        value="1"
      >
        second
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "minute",
          "value": 1,
        }
      }
    >
      <Option
        value="60"
      >
        minute
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "hour",
          "value": 1,
        }
      }
    >
      <Option
        value="3600"
      >
        hour
      </Option>
    </Localized>
  </withPropsOnChange(withUIContext(WithKeyboardFocus))>
</ForwardRef(forwardRef)>
`;

exports[`use initial unit if 0 1`] = `
<ForwardRef(forwardRef)
  itemGutter={true}
>
  <ForwardRef(forwardRef)
    aria-label="value"
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="DurationField-value"
    color="regular"
    disabled={false}
    name="duration-value"
    onChange={[Function]}
    spellCheck={false}
    textAlignCenter={true}
    value="0"
  />
  <withPropsOnChange(withUIContext(WithKeyboardFocus))
    aria-label="unit"
    classes={
      Object {
        "select": "DurationField-select",
      }
    }
    disabled={false}
    name="duration-unit"
    onChange={[Function]}
    value="1"
  >
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "second",
          "value": 0,
        }
      }
    >
      <Option
        value="1"
      >
        second
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "minute",
          "value": 0,
        }
      }
    >
      <Option
        value="60"
      >
        minute
      </Option>
    </Localized>
    <Localized
      id="framework-durationField-unit"
      vars={
        Object {
          "unit": "hour",
          "value": 0,
        }
      }
    >
      <Option
        value="3600"
      >
        hour
      </Option>
    </Localized>
  </withPropsOnChange(withUIContext(WithKeyboardFocus))>
</ForwardRef(forwardRef)>
`;
